dev-tools/omdb/src/bin/omdb/reconfigurator.rs

```diff
@@ -192,3 +224,125 @@ async fn cmd_reconfigurator_archive(
+    let mut targets: Vec<_> = {
+        use nexus_db_queries::db::schema::bp_target::dsl;
+        let conn = datastore
+            .pool_connection_for_tests()
+            .await
+            .context("obtaining connection")?;
+        dsl::bp_target
+            .select(BpTarget::as_select())
+            .order_by(dsl::version.desc())
+            .limit(i64::from(limit))
+            .get_results_async(&*conn)
+            .await
+            .context("listing targets")?
+    };
+
+    // Select everything from the blueprint table.
+    // This shouldn't be very large.
+    let mut all_blueprints: BTreeMap<BlueprintUuid, BlueprintMetadata> =
+        BTreeMap::new();
+    let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+    while let Some(p) = paginator.next() {
+        let records_batch = datastore
+            .blueprints_list(opctx, &p.current_pagparams())
+            .await
+            .context("batch of blueprints")?;
+        paginator = p.found_batch(&records_batch, &|b| *b.id.as_untyped_uuid());
+        all_blueprints.extend(records_batch.into_iter().map(|b| (b.id, b)));
+    }
+
+    // Sort the target list in increasing order.
+    // (This should be the same as reversing it.)
+    targets.sort_by_key(|b| b.version);
+
+    // Now, print the history.
+    println!("{:>5} {:24} {:36}", "VERSN", "TIME", "BLUEPRINT");
+    if targets.len() == usize::try_from(limit).unwrap() {
+        println!("... (earlier history omitted)");
+    }
+
+    let mut prev: Option<BlueprintUuid> = None;
+    let mut prev_blueprint: Option<Blueprint> = None;
+    for t in targets {
+        let target_id = BlueprintUuid::from(t.blueprint_id);
+        let this_previous = prev_blueprint.take();
+        prev_blueprint = None;
+
+        print!(
+            "{:>5} {} {} {:>8}",
+            t.version,
+            humantime::format_rfc3339_millis(t.time_made_target.into()),
+            target_id,
+            if t.enabled { "enabled" } else { "disabled" },
+        );
+
+        let same_blueprint = matches!(prev,
+            Some(prev_id) if prev_id == target_id
+        );
+        if same_blueprint {
+            // The only change here could be to the enable/disable bit.
+            // There's nothing else to say.
+            println!();
+        } else {
+            // The blueprint id changed.
+            let comment = match all_blueprints.get(&target_id) {
+                Some(b) => &b.comment,
+                None => "blueprint details no longer available",
+            };
+            println!(": {}", comment);
+
+            match (
+                history_args.diff,
+                prev.and_then(|prev_id| all_blueprints.get(&prev_id)),
+                all_blueprints.get(&target_id),
+            ) {
+                (true, Some(previous), Some(_)) => {
+                    let previous_blueprint = match this_previous {
+                        Some(p) => p,
+                        None => {
```

Comment:
How could we get to this branch? Shouldn't `this_previous` always be `Some(_)` if we've fetched the targets in order? (I guess we wouldn't have it on the very first blueprint we're showing, but diffing that against the preceding one we don't even list doesn't seem very important?)

Reply:
This is really non-obvious but this handles a case where the blueprint two iterations prior was missing from the database.  In that case, we won't have fetched the previous blueprint because _its_ parent was missing, so there was no point (we couldn't print a diff).

Another way to say it: `prev_blueprint` is only `Some` if: we're printing diffs, the previous blueprint's metadata is in the database, *and* we successfully printed its diff in the previous iteration, which means _its_ parent blueprint's metadata was in the database, too.

I made a few of the cleanups you suggested and added a bunch of comments here.